pkg/cri/resource-manager/cache/cache.go

```diff
@@ -645,6 +653,120 @@ func (cch *cache) GetConfig() *config.RawConfig {
 	return cch.Cfg
 }
 
+// SetAdjustment updates external adjustments and containers based on this.
+func (cch *cache) SetAdjustment(external *config.Adjustment) (bool, map[string]error) {
+	effective := map[*container]string{}
+
+	// collect per container external adjustments, checking for obvious errors
+	errors := map[string]error{}
+	for id, c := range cch.Containers {
+		if id != c.GetCacheID() {
+			continue
+		}
+
+		adjustments := cch.getApplicableAdjustments(external, c)
+
+		if len(adjustments) == 0 {
+			continue
+		}
+
+		// conflict: multiple adjustments per container
+		if len(adjustments) > 1 {
+			errors[c.GetID()] = cacheError("conflicting adjustments for %s: %s",
+				c.PrettyName(), strings.Join(adjustments, ","))
+			continue
+		}
+
+		adjust := external.Adjustments[adjustments[0]]
+
+		// error: trying to override resources for BestEffort container
+		if c.GetQOSClass() == v1.PodQOSBestEffort {
+			if adjust.Resources != nil {
+				errors[c.GetID()] = cacheError("%s: can't override resources for BestEffort %s",
+					adjustments[0], c.PrettyName())
+				continue
+			}
+		}
+
+		effective[c] = adjustments[0]
+	}
+	if len(errors) > 0 {
+		return false, errors
+	}
+
+	// update per container external adjustments, mark all containers with pending changes
+	for id, c := range cch.Containers {
+		if id != c.GetCacheID() {
+			continue
+		}
+
+		uptodate := effective[c]
+		previous := c.setEffectiveAdjustment(uptodate)
+		effective[c] = previous
+
+		if previous != uptodate {
+			cch.Info("%s effective external adjustment changed from %q to %q",
+				c.PrettyName(), previous, uptodate)
+		}
+
+		c.markPending(allControllers...)
+	}
+
+	if err := cch.Save(); err != nil {
+		for id, c := range cch.Containers {
+			if id != c.GetCacheID() {
+				continue
+			}
+			c.setEffectiveAdjustment(effective[c])
+		}
+		return false, map[string]error{"cache": err}
+	}
+
+	cch.External = external
```

Comment:
It's intentional that we use it until the agent gets its act together and notifies us about the current adjustments. This works identically for configuration. If someone changes the configuration while `cri-resmgr` is shut down, it will start up with the last cached version until the agent gets around to tell it otherwise.